controllers/vaultstaticsecret_controller.go

```diff
@@ -218,11 +245,242 @@ func (r *VaultStaticSecretReconciler) handleDeletion(ctx context.Context, o clie
+					"Error while watching events: %s", err)
+
+				if errorCount >= errorThreshold {
+					logger.Error(err, "Too many errors while watching events, requeuing")
+					break eventLoop
+				}
+
+				newVaultClient, err := r.ClientFactory.Get(ctx, r.Client, o)
+				if err != nil {
+					logger.Error(err, "Failed to retrieve Vault client")
+					break eventLoop
+				} else {
+					wsClient, err = newVaultClient.WebsocketClient(eventPath)
+					if err != nil {
+						logger.Error(err, "Failed to create new websocket client")
+						break eventLoop
+					}
+				}
+
+				// Update the LastClientID in the event registry
+				key := types.NamespacedName{Namespace: o.Namespace, Name: o.Name}
+				meta, ok := r.eventWatcherRegistry.Get(key)
+				if !ok {
+					logger.Error(
+						fmt.Errorf("failed to get event watcher metadata for VaultStaticSecret"),
+						"key", key.String())
+					break eventLoop
+				}
+				meta.LastClientID = newVaultClient.ID()
+				r.eventWatcherRegistry.Register(key, meta)
+			}
+		}
+	}
+
+	// If we've reached this point, we've encountered too many errors and need
+	// to close this watcher and requeue the resource
+	r.SourceCh <- event.GenericEvent{
+		Object: &secretsv1beta1.VaultStaticSecret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: o.Namespace,
+				Name:      o.Name,
+			},
+		},
+	}
+
+	return nil
+}
+
+// eventMsg is used to extract the relevant fields from an event message sent
+// from Vault
+type eventMsg struct {
+	Data struct {
+		Event struct {
+			Metadata struct {
+				Path     string `json:"path"`
+				Modified string `json:"modified"`
+			} `json:"metadata"`
+		} `json:"event"`
+		Namespace string `json:"namespace"`
+	} `json:"data"`
+}
+
+func (r *VaultStaticSecretReconciler) streamStaticSecretEvents(ctx context.Context, o *secretsv1beta1.VaultStaticSecret, wsClient *vault.WebsocketClient) error {
+	logger := log.FromContext(ctx)
+	conn, err := wsClient.Connect(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to connect to vault websocket: %w", err)
+	}
+	defer conn.Close(websocket.StatusNormalClosure, "closing event watcher")
+
+	// We made it past the initial websocket connection, so emit a "good" event
+	// status
+	r.Recorder.Eventf(o, corev1.EventTypeNormal, consts.ReasonEventWatcherStarted, "Started watching events")
+
+	for {
+		select {
+		case <-ctx.Done():
+			logger.V(consts.LogLevelDebug).Info("Context done, closing websocket",
+				"namespace", o.Namespace, "name", o.Name)
+			return nil
+		default:
+			msgType, message, err := conn.Read(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to read from websocket: %w, message: %q",
+					err, string(message))
+			}
+			messageMap := eventMsg{}
+			err = json.Unmarshal(message, &messageMap)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshal event message: %w", err)
+			}
+			logger.V(consts.LogLevelTrace).Info("Received message", "message type", msgType, "message", messageMap)
```

Comment:
Is it possible that `messageMap` may contain confidential info?

Reply:
I think we're pretty safe here since it's just the Path, Namespace, and Modified fields from the event message.